site/source/docs/porting/Debugging.rst

```diff
@@ -4,128 +4,200 @@
+When you want the fastest builds, you generally want to avoid generating large debug information
+during compile, because it takes time to link into the final binary. It is still worthwhile to use
+the ``-g2``
+flag (at link time only) because browsers understand the name section even when devtools are not 
+in use, resulting in more useful stack traces at minimal cost.
+
+Example:
 
 .. code-block:: bash
 
-  emcc -O1 -sASSERTIONS test/hello_world
+  emcc source.c -c -o source.o # source.o has no debug info
+  emcc source.o -o program.js -g2 # program.wasm has a name section, program.js is unminified
+
+Sometimes the use of the ``-O1`` or ``-Og`` flag at compile time can also result in faster
+builds, because optimizations early in the pipeline can reduce the amount of IR that is
+processed by later phases such as instruction selection and linking. It also of course
+reduces test runtime.
+
+
+Detecting Memory Errors and Undefined Behavior
+==============================================
+
+The best tools for detecting memory safety and undefined behavior issues. are Clang's sanitizers,
+such as the Undefined Behaviour Sanitizer (UBSan) and the Address Sanitizer (ASan).
+For more information, see :ref:`Sanitizers`.
+
+
+Emscripten has several other compiler settings that can be useful for catching errors at runtime.
+These are set using the :ref:`emcc -s<emcc-s-option-value>` option, and will override any optimization flags (TODO is this true?). For example:
+
+.. code-block:: bash
+
+  emcc -O1 -sASSERTIONS test/hello_world.c
 
 Some important settings are:
 
   -
     .. _debugging-ASSERTIONS:
 
-    ``ASSERTIONS=1`` is used to enable runtime checks for common memory allocation errors (e.g. writing more memory than was allocated). It also defines how Emscripten should handle errors in program flow. The value can be set to ``ASSERTIONS=2`` in order to run additional tests.
-
-    ``ASSERTIONS=1`` is enabled by default. Assertions are turned off for optimized code (:ref:`-O1 <emcc-O1>` and above).
+    ``ASSERTIONS=1`` is used to enable runtime checks for many types of common errors. It also
+    defines how Emscripten should handle errors in program flow. The value can be set to 
+    ``ASSERTIONS=2`` in order to run additional tests. ``ASSERTIONS=1`` is enabled by default at
+    ``-O0``.
 
   -
     .. _debugging-SAFE-HEAP:
 
-    ``SAFE_HEAP=1`` adds additional memory access checks, and will give clear errors for problems like dereferencing 0 and memory alignment issues.
-
-    You can also set ``SAFE_HEAP_LOG`` to log ``SAFE_HEAP`` operations.
+    ``SAFE_HEAP=1`` adds additional memory access checks with a Binaryen pass, and will give clear
+    errors for problems like dereferencing 0 and memory alignment issues.
+    You can also set ``SAFE_HEAP_LOG`` to log ``SAFE_HEAP`` operations. (TODO: any advantages over ASan?)
```

Comment:
@kripken IIRC ASan now supports sanitizing JS memory writes too, right? Is there any other reason a user might want to use SAFE_HEAP?